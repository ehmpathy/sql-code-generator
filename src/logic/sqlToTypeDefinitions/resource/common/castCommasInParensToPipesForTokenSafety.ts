export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
