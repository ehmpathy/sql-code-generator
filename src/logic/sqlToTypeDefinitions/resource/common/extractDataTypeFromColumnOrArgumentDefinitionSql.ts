import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
