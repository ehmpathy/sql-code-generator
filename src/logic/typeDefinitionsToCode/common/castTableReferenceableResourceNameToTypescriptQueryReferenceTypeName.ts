import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
