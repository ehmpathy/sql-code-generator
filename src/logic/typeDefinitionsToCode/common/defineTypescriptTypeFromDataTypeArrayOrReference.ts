import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
