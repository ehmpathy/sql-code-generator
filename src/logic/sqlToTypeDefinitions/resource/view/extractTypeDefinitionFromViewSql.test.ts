import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractTypeDefinitionFromViewSql } from './extractTypeDefinitionFromViewSql';

describe('extractTypeDefinitionFromViewSql', () => {
  it('should be able to extract types in this example', async () => {
    const exampleSql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/views/view_suggestion_current.sql`,
    });
    const typeDef = extractTypeDefinitionFromViewSql({ name: 'view_suggestion_current', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it('should be able to extract types in this other example', async () => {
    const exampleSql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/views/view_job_current.sql`,
    });
    const typeDef = extractTypeDefinitionFromViewSql({ name: 'view_suggestion_current', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it.todo('should throw an error if query has input variables');
});
