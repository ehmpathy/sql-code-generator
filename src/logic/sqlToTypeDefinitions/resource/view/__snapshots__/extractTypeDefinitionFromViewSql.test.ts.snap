// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTypeDefinitionFromViewSql should be able to extract types even when one of the columns has the word "from" in it 1`] = `
TypeDefinitionOfResourceView {
  "name": "view_email_current",
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "alias": "id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "uuid",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.uuid",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "external_id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.external_id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "ses_message_id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.ses_message_id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "to_address",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.to_address",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "from_address",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.from_address",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "from_name",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.from_name",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "subject",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.subject",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "body",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.body",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "status",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.status",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "created_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.created_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "effective_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.effective_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "updated_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.created_at",
      },
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "s",
      "tableName": "email",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "cvp",
      "tableName": "email_cvp",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "v",
      "tableName": "email_version",
    },
  ],
}
`;

exports[`extractTypeDefinitionFromViewSql should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceView {
  "name": "view_suggestion_current",
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "alias": "id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "uuid",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.uuid",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "suggestion_source",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.suggestion_source",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "external_id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.external_id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "status",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.status",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "result",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.result",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "created_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.created_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "effective_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.effective_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "updated_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.created_at",
      },
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "s",
      "tableName": "suggestion",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "cvp",
      "tableName": "suggestion_cvp",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "v",
      "tableName": "suggestion_version",
    },
  ],
}
`;

exports[`extractTypeDefinitionFromViewSql should be able to extract types in this other example 1`] = `
TypeDefinitionOfResourceView {
  "name": "view_job_current",
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "alias": "id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "uuid",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.uuid",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "title",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.title",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "status",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.status",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "description",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.description",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "location_id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.location_id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "photo_ids",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": "array_agg.output",
        "tableReferencePath": null,
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "created_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.created_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "effective_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.effective_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "updated_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.created_at",
      },
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "s",
      "tableName": "job",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "cvp",
      "tableName": "job_cvp",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "v",
      "tableName": "job_version",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "job_version_to_photo",
      "tableName": "job_version_to_photo",
    },
  ],
}
`;
